[Time Conductor] Reimplemented time conductor with new API and UI

akhenry opened the pull request:
This is an initial PR to start feedback. There are still a number of outstanding issues, but given the time constraints I'd like to start the code review ASAP.

### Description of changes
1. Introduces the time conductor public API into master for the first time
2. Defines new UI for the realtime/historical time conductor
3. Uses D3 for rendering the ticks. Introduces a new `MctAxis` directive to support this.
4. Introduces time systems, supported for now through a new legacy API extension point "TimeSystem", and supporting classes (eg. `TickSource`). Implements a default `UTCTimeSystem`, and a default `LocalClock`.
5. Central to the implementation is the `TimeConductorController`. Created some supporting classes around `TimeConductorController` such as `TimeConductorMode`. Possibly overkill here. Tried different variations with fewer class definitions, but this kept the code the cleanest even if it does result in a relatively large number of classes.

Note that after a discussion with @charlesh88 it was decided that it is better from a user perspective to select mode first, which will filter the available time systems. Time systems are filtered for real-time and LAD modes based on whether they define any appropriate tick sources.

Also note that although present in the UI, the zoom controller is non functional. It will be implemented if time permits, otherwise will be hidden via CSS.

Would suggest starting from the `TimeConductorController` and working outward from there.

### To do:
* [x] 1. Add support for formats to the D3 ticks. Currently uses the default D3 formatting, which imposes 12 hour time, and won't work for sols.
* [x] 2. Bring across telemetry decorator from old time conductor
* [x] 3. Need styles for snow for the time conductor v2 https://github.com/nasa/openmct/pull/1089#issuecomment-236656967 
* [x] 4. UTC time system should be a separate bundle from the conductor-v2, so that it can be disabled in other deployments https://github.com/nasa/openmct/pull/1089#issuecomment-236658245
* [x] 5. When switching to time systems that have a different format, the date time inputs do not update to use the correct format https://github.com/nasa/openmct/pull/1089#issuecomment-236674799
* [x] 6. Disable the zoom slider https://github.com/nasa/openmct/pull/1089#issuecomment-236732799 @charlesh88 
* [x] 7. Selecting a mode doesn't filter time systems https://github.com/nasa/openmct/pull/1089#issuecomment-236732799
* [x] 8. ~~Time systems without defaults totally break the time conductor~~ (deferred to #1091) https://github.com/nasa/openmct/pull/1089#issuecomment-236732878
* [x] 9. Tick generation fails when time system values are not dates
* [x] 10. ~~Ticks overlap when scale formatting is not available~~ (deferred to #1091) https://github.com/nasa/openmct/pull/1089#issuecomment-236732799
* [x] 11. Time conductor should be visible in edit mode and retain state on navigation https://github.com/nasa/openmct/pull/1089#issuecomment-237057156 and https://github.com/nasa/openmct/pull/1089#issuecomment-237057366.
* [x] 12. Specify delta format (needed for SCLK support)
* [x] 13. Unlisten to tick source on Time System change
* [x] 14. When a timesystem supports both LAD and Realtime, selecting realtime attaches a listener to the LAD clock instead of the realtime clock.
* [x] 15. Need to be able to set defaults per mode. a short hack would be to add mode: "fixed" or mode: "follow" to the Default and then pick the defaults that apply to the current mode.
* [x] 16. After navigating, unable to select time system .
* [x] 17. Changing mode sets time system to default even if previous time system was a valid choice.
* [x] 18. Better documentation (missing typedefs etc.)
* [x] 19. Lint and build errors
* [x] 20. Tests
* [x] 21. Squash commits

akhenry commented:
After discussion with @larkin two of the outstanding tasks above. Tracked in #1089.

larkin commented:
@akhenry note that we should add a issue to remove the compatibility layer / old functionality at a future date.

akhenry commented:
@larkin Have initial integration between d3 ticks and our `Format`s. It's a bit wonky at the moment because I'm basically re-implementing what D3 does, but not as well :) It's enough for code review though. So I believe everything is in place for the code review now.

akhenry commented:
@charlesh88 I've added status to the representation in plots. If you pull open933 you can play around with it. It's not perfect at the moment but should be enough to test your styling. 

To trigger the `s-status-timeconductor-unsynced` class

1. Open a plot (sine wave works well). __Class should not be applied at this point__
2. Select real-time mode in the time conductor
3. Marquee zoom in the plot. __Class should be applied at this point__

Optionally
4. Reset zoom, or switch out of real-time mode __Class should not be applied any more__

Just a caveat, the existing plot does not strictly observe fixed bounds. While changing the bounds in the conductor does set the bounds in the plot, the plot continues to tick as new data comes in.

larkin commented on platform/commonUI/formats/src/DurationFormat.js:
```diff
@@ -0,0 +1,60 @@
+
+    /**
+     * Formatter for UTC timestamps. Interprets numeric values as
+     * milliseconds since the start of 1970. Displays only the utc time. Can
```
is this jsdoc incorrect for the duration format?  Seems to be describing the handling for utcformat.

akhenry replied:
It's kind of correct, but also fairly confusing. I've tried to make it clearer in a new commit.

larkin commented on platform/features/conductor-v2/src/timeSystems/UTCTimeSystem.js:
```diff
@@ -0,0 +1,78 @@
+            'glyph': '\u0043'
+        };
+
+        //Time formats are defined as extensions. Include the key
```
Move this to a docstring on UTCTimeSystem.prototype.formats?  or just drop the comment entirely.

akhenry replied:
Dunzo

larkin commented on platform/features/conductor-v2/src/timeSystems/TickSource.js:
```diff
@@ -0,0 +1,54 @@
+
+define([], function () {
+    /**
+     * A tick source is a timing of timing signals. Usage is simple, a
```
"timing of timing signals?"

akhenry replied:
Replace with less nonsensical comment.

akhenry commented:
@larkin addressed those documentation issues.

larkin commented on platform/commonUI/formats/src/UTCTimeFormat.js:
```diff
@@ -34,6 +34,18 @@ define([
             "YYYY-MM-DD"
         ];
 
+    var MS = 1,
```
these constants are never used in this file.

akhenry replied:
Ah, good catch. They were used until very recently when I changed the implementation :)

larkin commented on platform/features/conductor-v2-compatibility/src/ConductorRepresenter.js:
```diff
@@ -0,0 +1,81 @@
+    [],
+    function () {
+
+        function ConductorRepresenter(
```
A tiny bit of jsdoc indicating this is transitional code would be great.  Don't need to document more than that.

larkin commented on platform/features/conductor-v2/src/TimeConductor.js:
```diff
@@ -0,0 +1,185 @@
+        } else if (bounds.start > bounds.end) {
+            return "Specified start date exceeds end bound";
+        }
+        return true;
```
Is `bounds.start === bounds.end` valid or invalid?

akhenry replied:
@larkin good question. I'm not sure? Perhaps if you knew the exact date and time that an event occurred you might use it to find the telemetry for that moment in time. Not so useful for plots, but might be useful in tables? There's no way in the design of directly entering a time of interest, so this could be a surrogate for that, as it would have the effect of setting the time of interest as well? It's a stretch though.

larkin replied:
In previous time conductor we said no, it's not valid.  I think it's a safe assumption to keep.  I think time of interest and/or annotations will handle the other cases you are mentioning above.

akhenry replied:
No problem, I'll change the logic to disallow bounds.start and bounds.end to be equal

larkin commented on platform/features/conductor-v2/src/TimeConductor.js:
```diff
@@ -0,0 +1,185 @@
+        return true;
+    };
+
+    function throwOnError(validationResult) {
```
function defined quite far from where it's used.  may be easier to inline.

akhenry replied:
Fair enough. Had intended to use it for validation elsewhere but at this stage there is only bounds validation. Have inlined for now. Can pull it out to a separate function if needed later.

larkin commented:
@akhenry I had the basic understanding that a time system would define a list of tick sources which the user would select from (e.g. LAD, realtime).  However, you've added another layer on top of this, modes, which are a method of mapping tick sources to the the user interface?  To someone defining custom time systems and tick sources this complexity appears to be hidden (I won't know for sure until I try and integrate it with VISTA), but from a maintainability perspective I'm wondering if it can be simplified?  I'm having trouble tracing the execution path for this code.

akhenry commented:
@larkin it's a fair point. There's really no provision in the UI design for selecting alternative tick sources, other than what is currently the 'mode' selector. Charles and I had a discussion about 'modes' and time systems and we felt that rather than the user selecting a time system and the UI filtering the tick sources/modes, it's a better user experience to select a 'mode', which will filter the supported time systems.

With that, mode becomes a first class concept again, so it made sense to implement it as such. A lot of the logic is mode specific, so it ended up being a nice way of encapsulating it as well. The code makes use of polymorphism quite a bit (it's hard to shake old habits :P ), which I think can serve to obfuscate code sometimes. That said, there ended up being a lot of classes in the implementation, so it might be possible to reduce that number, but potentially at the cost of complexity somewhere else.

akhenry commented:
@larkin have refactored a little and simplified modes somewhat. Pushes a little more code into the controller but I think it removes some of the mystery around modes. Also allowed me to remove two classes (LADMode.js and RealtimeMode.js).

larkin commented:
@akhenry @charlesh88 We need styles for snow for the time conductor v2.  Current look:

![screen shot 2016-08-01 at 10 57 23 am](https://cloud.githubusercontent.com/assets/226503/17303742/d22a7f9a-57d6-11e6-9140-e8fff14f021f.png)

charlesh88 commented:
@larkin @akhenry I think that's a merge problem - TC2 was looking fine in Snow in my topic branch. I'm in the latest open933 right now, will take a look.

larkin commented:
@akhenry I implemented a different format for a new time system.  When switching to time systems that have a different format, the date time inputs do not update to use the correct format, however their values are correctly updated from the defaults.

I manually removed the UTC time system from platform and created a new timesystem to verify that it's an issue with changing the time system and not necessarily a hardcoded default of UTC.

charlesh88 commented:
@akhenry @larkin Well, not sure what happenin' but wasn't a bad merge. Could have sworn this was looking fine in Snow, but on further inspection today realized I'd hard-coded some links to Espresso constants... Anyway, just pushed an updated version in branch open933-c that addresses this problem.

Andrew, my plan would be to keep going in this branch to do the work needed for #1076 - do you see any issue with that?

akhenry commented:
@charlesh88 Makes sense to keep it in one branch. I'll keep merging into open933

akhenry commented:
@larkin *interesting*. I'll admit I didn't try that. I'll take a look at it today.

akhenry commented:
@larkin will move UTCTimeFormat into another bundle while I'm at it.

akhenry commented:
@larkin @charlesh88  Have pushed a commit that reorganizes the bundles a little for the time conductor.

__old bundle structure__
```
- platform/features/conductor-v2
- platform/features/conductor-v2-compatibility
```
__new bundle structure__
```
- platform/features/conductor-v2/conductor
- platform/features/conductor-v2/compatibility
- platform/features/conductor-v2/utcTimeSystem
```
@larkin any code extending TimeSystem will need to reference its new location, which is `platform/features/conductor-v2/conductor/timeSystems`.

larkin commented:
> @akhenry https://github.com/nasa/openmct/pull/1089#issuecomment-235581764
> we felt that rather than the user selecting a time system and the UI filtering the tick sources/modes, it's a better user experience to select a 'mode', which will filter the supported time systems.

This doesn't actually occur; In my current configuration none of the time systems support realtime or LAD, and yet selecting real-time or lad still shows all available time systems.  Overall, not sure this is the right approach, from a user-discoverability point of view it requires them to understand the relationship between the two and provides no pathway for them to "figure it out" other than brute force.

akhenry commented:
@larkin hold off on pulling this for now. There are some issues.

larkin commented:
@akhenry ticks are happy to overlap when scale formatting is not available.  I thought this was prevented?  (not a blocker)

![screen shot 2016-08-01 at 3 39 51 pm](https://cloud.githubusercontent.com/assets/226503/17311208/56fba15a-57fe-11e6-9cb1-f445ca231236.png)

Also worth noting that d3 axis tick formatting only works with utc-like time systems.  Need a way to support it for other systems.  (also not a blocker).

larkin commented:
@akhenry how do I disable the zoom slider or give it different values?  No matter what time system I specify, it shows "Mars Minutes".

akhenry commented:
@larkin not prevented my d3 because we are handling the formatting ourselves in this case. Actually not confident it would handle it elegantly anyway in this case...

Long term answer is to adjust number of ticks based on label length. Short term answer might be to hard code a lower tick count, or use css ellipses

akhenry commented:
> @akhenry how do I disable the zoom slider or give it different values? No matter what time system I specify, it shows "Mars Minutes".

@charlesh88 put it in just in case we had time to get to it. Intention was to disable it via CSS if we didn't get to it in time. We haven't got to it in time :)

akhenry commented:
> This doesn't actually occur; In my current configuration none of the time systems support realtime or LAD, and yet selecting real-time or lad still shows all available time systems. Overall, not sure this is the right approach, from a user-discoverability point of view it requires them to understand the relationship between the two and provides no pathway for them to "figure it out" other than brute force.

Would be easy enough to disable or hide modes that are not supported by any available time systems.

larkin commented:
zoom slider: let's just disable for now.

tick overlap: Given our previous experience, CSS ellipses is a bad solution.  Ideally, we should read tick size from page and adjust number of ticks accordingly.  may result in a flicker but is only way that we can prevent overlapping ticks.  Let's defer this.

selecting a mode doesn't filter time systems: As I understand it, that's what you said happens or is intended to happen, but it doesn't currently happen in my deployment.  May just be incomplete?

larkin commented:
Something new: time systems without defaults totally break the time conductor.

I'm hardcoding some totally random default values for SCLK as a workaround.

charlesh88 commented:
Agree on the tick overlap: ellipses were always a lesser of the evils solution. Tick number should vary based on size of individual ticks, which should be based on the level of granularity of what needs to be displayed. With ticks basically useless right now in TC2 (until we get data visualization and/or Time of Interest) this is very deferable.

akhenry commented:
@larkin working on these issues, will have a fix for them all tonight

larkin commented:
@akhenry tick generation fails when time system values are not dates.  E.g. SCLK values are numbers, but sclkformat.format is being passed what appear to be date objects when generating ticks for the conductor axis.  So the values display as timestamps even though that is totally invalid.

charlesh88 commented:
@akhenry @larkin If we're pressed for time for JPL delivery, I say hide/disable the ticks temporarily.

akhenry commented:
> tick generation fails when time system values are not dates. E.g. SCLK values are numbers, but sclkformat.format is being passed what appear to be date objects when generating ticks for the conductor axis. So the values display as timestamps even though that is totally invalid.
@larkin 

Believe this is because the d3 axis is essentially in date mode, so it uses javascript dates as its internal representation of values. We can totally define non-date axes in D3. I need to do some analysis to see what the implications are of switching to a non-date D3 axis. Should be fine given that we are doing our own multi-scale tick formatting anyway.

akhenry commented:
Incidentally all, tracking outstanding issues in a checklist in the issue description.

akhenry commented:
Tracking outstanding issues in a checklist in the issue description.

@larkin feel free to rearrange them based on your priorities.

akhenry commented:
> tick generation fails when time system values are not dates. E.g. SCLK values are numbers, but sclkformat.format is being passed what appear to be date objects when generating ticks for the conductor axis. So the values display as timestamps even though that is totally invalid.
@larkin 

This is a tricky one. In short, I can pass number values to sclkformat.format, but the ticks intervals may  be unpredictable.

The long story is that, as you know, d3 is doing a lot of heavy lifting for us by providing a specific scale type for a time domain (rather than just a linear number domain). This affects how it formats the ticks (although we are overriding this), and also, crucially, the tick values that it generates. For dates, it uses a date-specific interpolation function that generates values that are whole time units (seconds, minutes, hours, days, etc. as appropriate) and represented as javascript date objects. 

I've made it so that tick formatting can be offloaded to a custom formatter, but the interpolation (and thus the calculation of tick values) is still being handled internally by d3 as though it is a time domain. Really, for SCLK, we need to hand off both to a custom class. Ticking on whole calendar values doesn't make any sense for SCLK. While a SCLK value can be wrapped in a Javascript date object, it's a meaningless representation, so the D3 interpolation will produce unpredictable ticks.

What we could do is use date interpolation if possible, but default to linear interpolation in all other cases. Only problem is, I don't know how to make that determination. The axis needs some sort of hint about how which interpolation method to use. This should probably be a property of the TimeSystem.

akhenry commented:
Have merged in latest from #1104.

larkin commented:
@akhenry can we discard #1104 as it has been merged here?

akhenry commented:
> @akhenry can we discard #1104 as it has been merged here?

👍

larkin commented:
@akhenry time conductor isn't visible in edit mode.

larkin commented:
@akhenry navigation resets time conductor to defaults instead of keeping the values entered by user.

akhenry commented:
> @akhenry navigation resets time conductor to defaults instead of keeping the values entered by user.

It does. @charlesh88 raised the same issue WRT mode. If you're in real-time mode, navigation shouldn't dump you into fixed mode.

The `TimeConductorController` needs a service layer to make it retain state between navigation. The `TimeConductor` is stateful, but the UI state cannot be completely reconstituted from it on controller initialization.

Mode aside, does it always make sense to retain state on navigation? The time conductor state may not be compatible with the newly navigated object. eg. navigating from MSL rover data in sols to JASON-3 data in SCET or ERT, or whatever. Maybe this doesn't matter though, it just displays nothing.

akhenry commented:
> @akhenry time conductor isn't visible in edit mode.

[It is now](https://github.com/nasa/openmct/pull/1089/commits/34c62ba4058db0085365ce3053481ffa27ea0926). But doesn't retain state

larkin commented:
> Mode aside, does it always make sense to retain state on navigation?

Yes.    Retaining state is vital for exploring telemetry.

MSL -> JASON3 Navigation is an extreme edge case.  I currently experience this in Demos and it's fine for now, easy to explain, users understand it.

akhenry commented:
@larkin Pushed refactored code that persists conductor state during navigation. Previously conductor was exposed directly as singleton angular service, but is now wrapped in TelemetryConductorService. I've updated the compatibility layer so hopefully you shouldn't need to make any code changes...

Although it works I want to do some refactoring as the code is a bit rough and ready at the moment.

akhenry commented:
Added a new toDate function to TimeSystem interface. Some time systems (eg SCLK) cannot meaningfully represent values as javascript dates. In this case, returning false from this function will prevent the time conductor from attempting to handle its values as dates. One consequence of this is that the time conductor will revert to using a linearScale instead of a utcScale.

larkin commented on platform/features/conductor-v2/conductor/src/ui/MctConductorAxis.js:
```diff
@@ -44,16 +44,27 @@ define(
+
+                    //If the time can be represented as a date, then use a
+                    // date scale. Otherwise use a standard linear scale
+                    if (timeSystem.toDate(bounds.start) !== false) {
```
You wrote a comment to explain unclear code.  Would be better to write code that is clear.

If the time system is utc based, then the only proper implementation of `toDate` is `new Date(val)`.  When describing the timesystem, describe what it is.

I suggest:
```javascript
if (timeSystem.isUTCBased()) { 
  xScale = d3.scaleUtc()
    .domain([new Date(start), new Date(end)]); 
} else { 
  xScale = d3.scaleLinear()
    .range([start, end]);
}
```

Using the return value of `toDate` for the check may be clever but is not clear.

larkin commented:
@akhenry few things I think we need:

1. need to be able to set defaults per mode.  a short hack would be to add `mode: "fixed"` or `mode: "follow"` to the Default and then pick the defaults that apply to the current mode.
2. once I have chosen LAD, I am unable to change the domain .
3. changing mode sets time system to default even if previous time system was a valid choice.

akhenry commented:
4. Specify duration format (needed for SCLK support)

larkin commented:
@akhenry when a timesystem supports both LAD and Realtime, selecting realtime attaches a listener to the LAD clock instead of the realtime clock.

larkin commented:
@akhenry No matter what I do the "duration" field never populates with a value unless I manually enter it, and when I do that it persists just fine.

![screen shot 2016-08-04 at 9 56 50 am](https://cloud.githubusercontent.com/assets/226503/17410762/fcd4183c-5a29-11e6-8616-991f67f9e23e.png)

larkin commented:
@akhenry I think this may help demonstrate what's happening, captured from a fresh load.
![mode-switch-buggy](https://cloud.githubusercontent.com/assets/226503/17410988/033fdf98-5a2b-11e6-9fca-77b78258f3e5.gif)

1. fixed defaults don't include defaults they don't make sense in fixed would they?
2. defaults aren't reapplied on change of mode, thus when I specify a duration and switch back the fixed mode shows me the same thing as realtime was previously showing.

akhenry commented:
> 1.fixed defaults don't include defaults they don't make sense in fixed would they?

Assuming you meant deltas here, in which case you're correct. If you specify deltas in the defaults for fixed mode they would be ignored.

akhenry commented:
@larkin pushed a fix for https://github.com/nasa/openmct/pull/1089#issuecomment-237618402

akhenry commented:
Merged from master to pickup popup menu fix to check for regression in new time conductor.

akhenry commented:
@larkin I've done some refactoring, and added tests. It's ready for a final review and merge.

The bulk of the refactoring was creating a new `TimeConductorViewService` which moved some of the responsibilities away from the controller. The controller now just populates the view, and responds to user input. With a little extra code the new view service could be used by other code to set mode and deltas on the time conductor.

I've also simplified Mode related code, and consolidated everything into one Mode class. The differences between the modes are now simply defined in metadata and tick sources.

The main difference that will affect VISTA is that I've removed the 'type' function from tick sources. It's replaced with a 'mode' specifier on the tick source metadata. The options are 'realtime' and 'LAD'. This replaces 'clock' and 'data' which required an extra step to map them to modes.

akhenry commented:
Resolved merge conflicts

akhenry commented:
Merged from master to resolve build issue

larkin commented:
# Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

larkin commented:
Squash merged in an integration branch, closing as merged.
